core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.InstancedConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.ReadRequest;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.LockResource;
+import alluxio.util.ConfigurationUtils;
+import alluxio.wire.WorkerNetAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A shared gRPC data reader that cache blocks data for multi-thread accessing.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public class SharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(SharedGrpcDataReader.class);
+  private static final ReentrantReadWriteLock[] BLOCK_LOCKS;
+  /** A map from block id to the block's cached data reader. */
+  @GuardedBy("BLOCK_LOCKS")
+  private static final ConcurrentHashMap<Long, BufferCachingGrpcDataReader> BLOCK_READERS
+      = new ConcurrentHashMap<>();
+
+  static {
+    AlluxioConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+    int lockNum = conf.getInt(PropertyKey.FUSE_SHARED_CACHING_READER_LOCKS);
```

Comment:
I don't think we need another flag `FUSE_SHARED_CACHING_READER_LOCKS` on this.
Picking a constant number like 32 is good enough to ensure ~32 concurrent readers.

In general, we should be very conservative in adding new configuration flags.
Unless we're very sure users will tune it.

Reply:
Yeah, it may not be helpful and necessary to add it as a property, removed